### DISCUSSION TITLE
[Question][Hw3]

### DISCUSSION THREAD

**Discussion by ee12ha0220** [prosyslab-classroom/cs348-information-security#517] [Q&A]
What does "clang (filename).c (filename2).ll" do?

**Comment by RiceBiscuits** [comment]
Specified files (filename).c and (filename2.)ll will be compiled together.
If you want to know more thing about Clang, please check the link below.
(https://clang.llvm.org/docs/UsersManual.html)

**Comment by ee12ha0220** [comment]
I have searched the povided link, but there is no information about how .ll file is compiled by clang. Can you explain this part?

**Comment by KihongHeo** [comment]
Please take a look at https://github.com/prosyslab-classroom/llvm-primer.

**Comment by ee12ha0220** [comment]
I have checked the link already, but there is only information about creating .ll file using clang, not compiling .ll file with clang

**Comment by KihongHeo** [comment]
`clang` is a C compiler.
We use `clang` to create ll files from c files. The internal details of compilation are not a part of this homework.
Could you tell me why you need that one?

**Comment by ee12ha0220** [comment]
In the Makefile in the test folder, there is a line "$(CC) -o $* runtime.c $* .instrumented.ll". 
So I was wondering what effect will "$* .instrumented.ll" give in compiling "runtime.c" with clang, in this situation.

**Comment by KihongHeo** [comment]
- `runtime.c` contains the runtime infrastructure for sanitizer
- `X.ll` is an intermediate representation (IR) of some `X.c`
- `X.instrumented.ll` should be the same as `X.ll` but with more instrumentation that uses functions in `runtime.c`
- `clang X.instrumented.ll runtime.c` will generate an instrumented executable file. The behavior of the exe file will be the same as that from `clang X.ll` or from `clang X.c`. But it has additional sanitization and  linked with the sanitization functions.

**Comment by ee12ha0220** [comment]
Thank you :)
